### PR TITLE
[INFRA] new compilers

### PR DESCRIPTION
**Pull request by eseiler**
This adds CI for recent compilers. We still need to decide on the build matrix.

### CODE REVIEW

File: include/seqan/basic/debug_test_system.h [hunk 1]
```diff
@@ -164,7 +164,18 @@
...
+// gcc 9.3 -std=c++2a has ignored attribute warnings on macOS
+#if defined(__GNUC__) && defined(__APPLE__) && (__GNUC__ == 9) && __cplusplus > 201703L
+#define unused_backup __unused 
+#undef __unused
+#define __unused 
 #include <dirent.h>     // DIR
+#undef __unused
+#define __unused unused_backup
+#undef unused_backup
+#else
```
Comment (marehr) [inline include/seqan/basic/debug_test_system.h:176]:
```suggestion
#else // #if defined(__GNUC__) && defined(__APPLE__) && (__GNUC__ == 9) && __cplusplus > 201703L
```

File: include/seqan/basic/debug_test_system.h [hunk 2]
```diff
@@ -164,7 +164,18 @@
...
+#include <dirent.h>     // DIR
+#endif
```
Comment (marehr) [inline include/seqan/basic/debug_test_system.h:178]:
```suggestion
#endif // #if defined(__GNUC__) && defined(__APPLE__) && (__GNUC__ == 9) && __cplusplus > 201703L
```

File: include/seqan/basic/debug_test_system.h [hunk 3]
```diff
@@ -164,7 +164,18 @@
...
 #include <sys/stat.h>   // mkdir()
+// gcc 9.3 -std=c++2a has ignored attribute warnings on macOS
+#if defined(__GNUC__) && defined(__APPLE__) && (__GNUC__ == 9) && __cplusplus > 201703L
+#define unused_backup __unused 
+#undef __unused
+#define __unused 
 #include <dirent.h>     // DIR
+#undef __unused
+#define __unused unused_backup
+#undef unused_backup
```
Comment (marehr) [inline include/seqan/basic/debug_test_system.h:175]: this does not seem to work:

https://godbolt.org/z/75aT9M

I found this answer:
https://stackoverflow.com/a/1545079

This seems to work:
https://godbolt.org/z/TMd3f9

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -148,16 +166,16 @@ jobs:
           if [ "$RUNNER_OS" == "Linux" ]; then
             sudo apt-get install --yes ${{ matrix.cxx }}
           else
-            if [ "${{ matrix.cxx }}" == "clang++-7" ]; then
-              cxx_macos_name="llvm@7"
+            if [[ "${{ matrix.cxx }}" == "clang++-"* ]]; then
+              cxx_macos_name=$(echo "${{ matrix.cxx }}" | sed "s/clang++-/llvm@/g")
```
Comment (marehr) [inline .github/workflows/ci.yml:170]:
```suggestion
              cxx_macos_name=$(echo "${{ matrix.cxx }}" | sed "s/clang++-/llvm@/g") # e.g. llvm@7
```

**Review by marehr — CHANGES_REQUESTED** [review]
Thank you for doing this :) Just some small remarls

File: include/seqan/basic/debug_test_system.h (hunk 3 above)
Comment (eseiler) [reply include/seqan/basic/debug_test_system.h]: It somehow worked, even though it shouldn't :)

**Comment by eseiler** [comment]
The last open thing would be to decide on the build matrix (gcc 8,9,10, (11?), clang 8,9,10, std17, std20, ...).

**Review by marehr — APPROVED** [review]
Thank you!

**Comment by marehr** [comment]
> The last open thing would be to decide on the build matrix (gcc 8,9,10, (11?), clang 9,10,11 std17, std20, ...).

We should test the highest supported standard for each compiler.

**Comment by eseiler** [comment]
So,
* gcc 8 -std=c++17
* gcc 9 -std=c++2a
* gcc 10 - std=c++20
* clang 9 -std=c++2a
* clang 10 -std=c++20
* clang 11 -std=c++20
* MSVC /std:c++latest (/std:c++20 will only be available once it is fully implemented)
* then?

And then drop 8 and test 11 once it is out

**Comment by marehr** [comment]
> That means:
> 
>     * gcc 8 -std=c++17
> 
>     * gcc 9 -std=c++2a
> 
>     * gcc 10 - std=c++20
> 
>     * clang 9 -std=c++2a
> 
>     * clang 10 -std=c++20
> 
>     * clang 11 -std=c++20
> 
>     * MSVC /std:c++latest (/std:c++20 will only be available once it is fully implemented)
> 
> 
> And then drop gcc 8 and test gcc 11 once it is out.
> Is this ok?

Thank you for spelling that out. If I see the list like that, I'm wondering if we should stick to the highest stable standard.

* gcc 8 -std=c++17
* gcc 9 -std=c++17
* gcc 10 - std=c++20
* clang 9 -std=c++17
* clang 10 -std=c++20
* clang 11 -std=c++20
* MSVC /std:c++latest (/std:c++20 will only be available once it is fully implemented)

**Comment by eseiler** [comment]
Then I would use clang 9 and gcc 9with both std=c++17 and c++2a. The macro thingie and `seqan::erase` only occur in std=c++2a.

**Comment by marehr** [comment]
Okay :)

**Comment by eseiler** [comment]
I forgot to add the cxx_flags for the std=c++20 stuff, so some stuff will fail :D
There is no brew for llvm@10 because security/maintanence reasons; we can either tap some 3rd party repo, or just remove it (clang10 macOS).

**Comment by marehr** [comment]
If there is no official package, yet, don't test it ;)

**Comment by eseiler** [comment]
> If there is no official package, yet, don't test it ;)

There apparently won't be. There is llvm 7,8,9,11
I'll remove it

**Comment by marehr** [comment]
> > If there is no official package, yet, don't test it ;)
> 
> There apparently won't be. There is llvm 7,8,9,11
> I'll remove it

Interesting?! Kay!

File: include/seqan/stream/stream_base.h [hunk 5]
```diff
@@ -206,7 +206,7 @@ guessFormatFromStream(TStream &istream, Tag<TFormat_>)
 
     // check magic header
     unsigned i;
-    for (i = 0; i != sizeof(MagicHeader<TFormat>::VALUE) / sizeof(char); ++i)
+    for (i = 0; i != sizeof(MagicHeader<TFormat>::VALUE) / (sizeof(char)); ++i)
```
Comment (marehr) [inline include/seqan/stream/stream_base.h:209]: [FIX] suppress Wsizeof-pointer-div

Hä? What?

File: include/seqan/basic/basic_type.h [hunk 6]
```diff
@@ -499,7 +499,7 @@ _referenceCast(typename RemovePointer<T>::Type & ptr)
 template <typename T>
 struct LENGTH
 {
-    enum { VALUE = 1 };
+    static const unsigned VALUE = 1;
```
Comment (marehr) [inline include/seqan/basic/basic_type.h:502]: I guess there are multiple different LENGTH metafunctions

File: apps/fiona/CMakeLists.txt [hunk 7]
```diff
@@ -65,6 +65,10 @@ add_executable (fiona_illumina fiona.cpp)
 
 set_target_properties (fiona_illumina PROPERTIES COMPILE_DEFINITIONS "FIONA_ILLUMINA")
 
+# Boost 1.74.0 is not compatible with cpp20
```
Comment (marehr) [inline apps/fiona/CMakeLists.txt:68]: Could we upgrade to a newer version later?

File: include/seqan/map/map_base.h [hunk 8]
```diff
@@ -43,13 +43,13 @@ namespace seqan
 // Forwards
 // ==========================================================================
 
-#if !defined(_MSC_VER)
+#if !defined(_MSC_VER) || (defined(_MSVC_LANG) && _MSVC_LANG >= 201703)
```
Comment (marehr) [inline include/seqan/map/map_base.h:46]: 👍🏽

**Review by marehr — COMMENTED** [review]
Some questions

File: include/seqan/stream/stream_base.h (hunk 5 above)
Comment (eseiler) [reply include/seqan/stream/stream_base.h:209]: There is a warning when you do this kind of division. It's supposed to warn if you try to get the size of an array by dividing the sizeof the array by the sizeof a pointer to the array. You can silence it with parentheses...

File: apps/fiona/CMakeLists.txt (hunk 7 above)
Comment (eseiler) [reply apps/fiona/CMakeLists.txt:68]: We could, but there is no compatible version yet.

File: include/seqan/basic/basic_type.h (hunk 6 above)
Comment (eseiler) [reply include/seqan/basic/basic_type.h:502]: I.e. I should double-check if there are different function that still use the enums?

File: include/seqan/stream/stream_base.h (hunk 5 above)
Comment (marehr) [reply include/seqan/stream/stream_base.h:209]: Interesting, do you know why this is in general bad, why this is okay here?

File: include/seqan/basic/basic_type.h (hunk 6 above)
Comment (marehr) [reply include/seqan/basic/basic_type.h:502]: Wo kein Ankläger, da ist kein Angeklagter. Not blocking for me.

File: apps/fiona/CMakeLists.txt (hunk 7 above)
Comment (marehr) [reply apps/fiona/CMakeLists.txt:68]: Ah okay, so that might be a move in the future :)

**Review by marehr — APPROVED** [review]

File: include/seqan/stream/stream_base.h (hunk 5 above)
Comment (eseiler) [reply include/seqan/stream/stream_base.h:209]: It's only a problem if the array is passed as function parameter (which we are not doing here).

from the patch to gcc (https://patchwork.ozlabs.org/project/gcc/patch/AM4PR0701MB2162D22EC7A12EB0797E231DE41B0@AM4PR0701MB2162.eurprd07.prod.outlook.com/):
```cpp
int
f1 (int *array)
{
  int i;
  i = sizeof array / sizeof *array;		/* { dg-warning "dividing the pointer size by the element size" } */
  i += sizeof array / sizeof array[0];		/* { dg-warning "dividing the pointer size by the element size" } */
  i += sizeof(array) / sizeof(*array);		/* { dg-warning "dividing the pointer size by the element size" } */
  i += sizeof(array) / sizeof(array[0]);	/* { dg-warning "dividing the pointer size by the element size" } */
  i += (sizeof(array)) / (sizeof(array[0]));	/* { dg-warning "dividing the pointer size by the element size" } */
  i += sizeof(array) / sizeof(int);		/* { dg-warning "dividing the pointer size by the element size" } */
  i += sizeof(array) / sizeof(char);
  i += sizeof(*array) / sizeof(char);
  i += sizeof(array[0]) / sizeof(char);
  return i;
}

int
f2 (void)
{
  int array[10];
  int i;
  i = sizeof array / sizeof *array;
  i += sizeof array / sizeof array[0];
  i += sizeof(array) / sizeof(*array);
  i += sizeof(array) / sizeof(array[0]);
  i += (sizeof(array)) / (sizeof(array[0]));
  i += sizeof(array) / sizeof(int);
  i += sizeof(array) / sizeof(char);
  i += sizeof(*array) / sizeof(char);
  i += sizeof(array[0]) / sizeof(char);
  return i;
}
```

**Pull request merged**